src/v/storage/mvlog/skipping_data_source.cc

```diff
@@ -0,0 +1,58 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "storage/mvlog/skipping_data_source.h"
+
+#include "io/pager.h"
+#include "io/paging_data_source.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/temporary_buffer.hh>
+
+using ::experimental::io::paging_data_source;
+
+namespace storage::experimental::mvlog {
+
+ss::future<ss::temporary_buffer<char>> skipping_data_source::get() noexcept {
+    while (true) {
+        // Initialize a stream if we haven't already or we've exhausted the
+        // previous one.
+        if (!cur_stream_) {
+            if (reads_.empty()) {
+                co_return ss::temporary_buffer<char>();
+            }
+            auto first_read = *reads_.begin();
+            const auto end_pos = pager_->size();
+            if (first_read.offset >= end_pos) {
+                // Skip the read if starts past the end of the file.
+                reads_.pop_front();
```

Comment:
when does this happen?

Reply:
As of now it's not expected -- callers will be expected to not fetch past the end. But it's an easy enough case to support. I can remove it and replace with an assert if you prefer a leaner set of requirements.